src/nc_rbtree.c

```diff
@@ -342,3 +342,44 @@ rbtree_delete(struct rbtree *tree, struct rbnode *node)
 
     rbtree_black(temp);
 }
+
+struct rbnode *rbtree_find(struct rbtree *tree, int64_t key) {
+    struct rbnode **root = &tree->root;
+    struct rbnode *sentinel = tree->sentinel;
+    struct rbnode *temp, **p;
+
+    /* empty tree */
+
+    if (*root == sentinel) {
+        return NULL;
+    }
+
+    /* a binary tree find */
+    temp = *root;
+    for (;;) {
+
+        if (temp->key == key) {
+            break;
+        }
+
+        p = (key < temp->key) ? &temp->left : &temp->right;
+        if (*p == sentinel) {
+            return NULL;
+        }
+        temp = *p;
+    }
+
+    return temp;
+}
+
+void rbtree_inorder_traversal(struct rbnode *root, struct rbnode *sentinel, 
```

Comment:
yes, it`s a pre-order traversal ! Thanks!

Reply:
By compare, i think use STALQ for conn list is better than `struct array`; because the `struct array` delete elem will trigger chain copy.

Reply:
Either STAILQ or struct array is fine.

STAILQ is also fine since I typically expect at most one monitor at a time and checking if the head is non-null is also efficient

Additions and removals should be so infrequent (and the typical number of running monitors) that the chain copy wouldn't matter - if you have dozens of different monitors running at once you have much bigger performance problems

A chain copy can also be avoided in this case by moving the last element into the removed element's position and decrementing the size